mlir/lib/Dialect/Rock/Transforms/GemmOutputSwizzle.cpp

```diff
@@ -0,0 +1,571 @@
+                                   /*extraViews=*/idToLDS,
+                                   /*extraIndices=*/ValueRange{tid},
+                                   op.getFeatures(), StoreMethod::Set,
+                                   /*forceUnroll=*/forceUnroll,
+                                   /*useIndexDiffs=*/useIndexDiffs);
+
+    // Load from LDS to registers.
+    int64_t elementsWrittenPerThread = 128 / destType.getIntOrFloatBitWidth();
+    elementsWrittenPerThread =
+        math_util::gcd(dataPerThread, elementsWrittenPerThread);
+    int64_t iter = dataPerThread / elementsWrittenPerThread;
+    LLVM_DEBUG(llvm::dbgs()
+               << "blockSize: " << blockSize
+               << " dataPerThread: " << dataPerThread
+               << " elementsWrittenPerThread: " << elementsWrittenPerThread
+               << " iter: " << iter << "\n");
+    if (dim == dimensionM) {
+      LLVM_DEBUG(llvm::dbgs() << "dim = M\n");
+    } else {
+      LLVM_DEBUG(llvm::dbgs() << "dim = N\n");
+    }
+
+    Value finalC =
+        gpuAlloc(b, loc, dataPerThread, destType, AddressSpace::Private);
+
+    TopDownTMBuilder tidIterMerge(b, {"tid", "iter"},
+                                  {blockSize, dataPerThread});
+    tidIterMerge.passThrough(ArrayRef<StringRef>{"tid"});
+    tidIterMerge.merge({"iter", "numElements"}, {1, 2}, "iter",
+                       {iter, elementsWrittenPerThread});
+    auto tidIterMergeAttr = tidIterMerge.get();
+
+    auto tidIterFlatten =
+        TopDownTMBuilder::below(tidIterMerge, tidIterMergeAttr);
+    tidIterFlatten.unmerge("flattenBlock", 0, {"iter", "tid", "numElements"},
+                           {iter, blockSize, elementsWrittenPerThread});
+    auto tidIterFlattenAttr = tidIterFlatten.get();
+
+    SmallVector<Attribute> transformAttrs;
+    transformAttrs.push_back(tidIterMergeAttr);
+    transformAttrs.push_back(tidIterFlattenAttr);
+
+    ArrayAttr ldsRead = b.getArrayAttr(transformAttrs);
+    auto ldsBufferForLoad = transform(b, typedBuffer, ldsRead);
+
+    // LDS barrier.
+    b.create<LDSBarrierOp>(loc);
+
+    b.create<ThreadwiseReadIntoOp>(loc, ldsBufferForLoad, finalC,
+                                   b.getArrayAttr({}), ValueRange{tid},
+                                   forceUnroll, useIndexDiffs);
+
+    SmallVector<int64_t, 5> bidGridLengths;
+    SmallVector<StringRef, 5> bidGridOrder;
+    int64_t mBlocks = M / mPerBlock;
+    int64_t nBlocks = N / nPerBlock;
+    if (M % mPerBlock != 0 || N % nPerBlock != 0) {
+      return failure();
+    }
+    if ((mBlocks * nBlocks) != gridSize) {
+      return failure();
+    }
+    bool isAttention = op.getExtraIndices().size() == 3;
+    LLVM_DEBUG(llvm::dbgs() << "isAttention: " << isAttention << "\n");
+    if (isAttention) {
+      // attention
+      dataPerThread = dataPerThread * mBlocks;
+      mBlocks = 1;
+      bidGridLengths = {G, nBlocks, blockSize, dataPerThread};
+      bidGridOrder = {"g_block", "n_block", "tid", "iter"};
```

Comment:
These are more than grid dimensions. Therefore I d not name like "bidGrid*".

These are upperView of threadwise_write_all. Therefore, I dont see the requirement to construct it like this.
Can we just extract them instead?

Reply:
Isn't the point here to reconstruct the view to account for the swizzles?